### ISSUE TITLE
Warning on redundant trait bounds

### ISSUE THREAD

**Issue by ltrzesniewski** [rust-lang/rust#102485]
This has initially been reported by @jruderman [on Twitter](https://twitter.com/jruderman/status/1572866237914558465?s=20&t=hHIhvA6m1SMOjcRb930Lyw) (cc @estebank who replied there).

Currently, rustc does not generate any warning for redundant trait bounds.

Given the following code: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=00cd7a97f81ad47d7502604d55df5759)

```rust
use std::fmt::Display;

fn main() {
    foo::<i32>();
}

fn foo<T: Display + Display>()
where
    T: Display + Display,
    T: Display,
{
}
```

The current output is:

```
   Compiling rust-temp v0.1.0 (C:\Dev\Projects\rust-temp)
    Finished dev [unoptimized + debuginfo] target(s) in 0.33s
```

So I guess the questions are:

- Should redundant trait bounds produce warnings?
- Should supertraits be considered? (I get the feeling this could generate lots of warnings in existing code, or make refactoring more cumbersome)
- Should this go into rustc or clippy?

If you think this would be a good fit for a first contribution to rustc, I'd like to give it a try. 🙂 

@rustbot claim

**Comment by compiler-errors** [comment]
I feel like this should be a clippy lint if anything -- redundant bounds could, for example, be generated by macros or other sorts of situations, and they're really not that dangerous.

**Comment by estebank** [comment]
@compiler-errors we do already silence lints coming from macros. I wouldn't mind if this was a clippy lint though.

**Comment by RodBurman** [comment]
As of 1.82.0
% cargo -V -v
cargo 1.82.0 (8f40fc59f 2024-08-21)
release: 1.82.0
commit-hash: 8f40fc59fb0c8df91c97405785197f3c630304ea
commit-date: 2024-08-21
host: aarch64-apple-darwin
libgit2: 1.8.1 (sys:0.19.0 vendored)
libcurl: 8.7.1 (sys:0.4.74+curl-8.9.0 system ssl:(SecureTransport) LibreSSL/3.3.6)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: Mac OS 15.1.0 [64-bit]

The above code gives:
% cargo build
   Compiling redundanttb v0.1.0 (/Users/rod/code/rust/triage/redundanttb)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.37s

And:
% cargo clippy
    Checking redundanttb v0.1.0 (/Users/rod/code/rust/triage/redundanttb)
warning: bound is defined in more than one place
 --> src/main.rs:7:8
  |
7 | fn foo<T: Display + Display>()
  |        ^
8 | where
9 |     T: Display + Display,
  |     ^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_bound_locations
  = note: `#[warn(clippy::multiple_bound_locations)]` on by default

warning: bound is defined in more than one place
  --> src/main.rs:7:8
   |
7  | fn foo<T: Display + Display>()
   |        ^
...
10 |     T: Display,
   |     ^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_bound_locations

warning: `redundanttb` (bin "redundanttb") generated 2 warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.11s

So I think the clippy lint is now included and we can close this issue?

**Comment by ltrzesniewski** [comment]
Oh, sorry, I forgot to close this after the team members said it should be handled by clippy instead.
Happy to see it has been implemented! 👍 

@rustbot release-assignment
